### PR TITLE
[Neurips23] ParlayANN Submission for OOD track

### PR DESCRIPTION
**Pull request by magdalendobson**
This PR contains our submission for the OOD track. We tested on an AWS c6i.2xlarge and found that at 90% recall, our entry was averaging around 7000-7500 QPS, while the baseline was around 3500. The build took about 4 hours. I found that on this machine, when building and then querying, the main memory did not seem to flush out between saving the files and then reloading them and thus overran the memory limit. This means that it was necessary to build the index and then load and query it on a separate run. I observed the same problem with the DiskANN baseline. Just wanted to note that in case it comes up during evaluation.

I have edited the neurips23.yml file to test our build using the random-xs dataset. Please let me know if there is anything else needed.

### CODE REVIEW

File: .github/workflows/neurips23.yml [hunk 1]
```diff
@@ -21,6 +21,9 @@ jobs:
           - algorithm: diskann
             dataset: random-xs
             track: ood
+          - algorithm: vamana
```
Comment (harsha-simhadri) [inline .github/workflows/neurips23.yml:24]: please check indentation

File: benchmark/dataset_io.py [hunk 2]
```diff
@@ -16,6 +16,8 @@
 def download(src, dst=None, max_size=None):
     """ download an URL, possibly cropped """
     if os.path.exists(dst):
+        print(dst)
```
Comment (harsha-simhadri) [inline benchmark/dataset_io.py:19]: please dont modify the benchmark folder.

File: benchmark/runner.py [hunk 3]
```diff
@@ -290,8 +290,8 @@ def run_docker(definition, dataset, count, runs, timeout, rebuild,
  
     # set/override container timeout based on competition flag
     if neurips23track!='none':
-        timeout = 30*60 # 30 minutes
-        print("Setting container wait timeout to 30 minutes")       
+        timeout = 12*60*60 # 12 hours
```
Comment (harsha-simhadri) [inline benchmark/runner.py:293]: please don't modify this

**Review by harsha-simhadri — CHANGES_REQUESTED** [review]
Thanks for the heads up on memory. Will try out your PR.

File: neurips23/ood/diskann/Dockerfile [hunk 4]
```diff
@@ -10,5 +10,5 @@ RUN git clone https://github.com/microsoft/DiskANN.git --branch  0.5.0.rc3
 WORKDIR /home/app/DiskANN
 RUN pip3 install virtualenv build
 RUN python3 -m build
-RUN pip install dist/diskannpy-0.5.0rc3-cp310-cp310-linux_x86_64.whl
+RUN pip install dist/diskannpy-0.5.0rc2-cp310-cp310-linux_x86_64.whl
```
Comment (harsha-simhadri) [inline neurips23/ood/diskann/Dockerfile:13]: please remove this change from your PR

**Comment by harsha-simhadri** [comment]
I am unable to run this since the time out for docker is set to 30 minutes in your branch. You may want to rebase with the latest commit on this repo's main to get the latest runner config.

**Comment by magdalendobson** [comment]
> I am unable to run this since the time out for docker is set to 30 minutes in your branch. You may want to rebase with the latest commit on this repo's main to get the latest runner config.

I just pulled from main and rebased.

**Comment by magdalendobson** [comment]
Hmm, yeah, I think it's possible. Let me experiment with modifying the delete constructors of some of my objects and I'll get back to you.

**Comment by magdalendobson** [comment]
@maumueller thanks for the suggestion. I took a closer look and it did turn out that some of my data structures were not freeing memory properly. You should now be able to re-install and then run my submission without memory problems.

**Comment by harsha-simhadri** [comment]
I see the following results after restarting from a crash post index build

vamana,vamana,text2image-10M,10,8568.28073422188,0.0,1000000.0,13098176.0,1528.6819382196045,0,0,ood,0.8773770000000001
vamana,vamana,text2image-10M,10,7777.599618424573,0.0,1000000.0,13098176.0,1684.089776101532,0,0,ood,0.88975
vamana,vamana,text2image-10M,10,7128.303182660356,0.0,1000000.0,13098176.0,1837.488623079529,0,0,ood,0.8999010000000001
vamana,vamana,text2image-10M,10,6873.170646655227,0.0,1000000.0,13098176.0,1905.6963188269044,0,0,ood,0.9041650000000001
vamana,vamana,text2image-10M,10,6607.938732179394,0.0,1000000.0,13098176.0,1982.1878699047852,0,0,ood,0.908165
vamana,vamana,text2image-10M,10,6376.8978543202775,0.0,1000000.0,13098176.0,2054.0043606196596,0,0,ood,0.911952
vamana,vamana,text2image-10M,10,6148.750960929249,0.0,1000000.0,13098176.0,2130.2173536103824,0,0,ood,0.9153420000000001
vamana,vamana,text2image-10M,10,5739.398208736937,0.0,1000000.0,13098176.0,2282.1514597229,0,0,ood,0.92154
vamana,vamana,text2image-10M,10,5551.22053214029,0.0,1000000.0,13098176.0,2359.512817796478,0,0,ood,0.924339
vamana,vamana,text2image-10M,10,4789.167993183752,0.0,1000000.0,13098176.0,2734.958560368347,0,0,ood,0.935529

**Comment by magdalendobson** [comment]
That seems about right. Are you sure you pulled the latest version if you're still having the crash issue? My latest push fixed it on the 16G machine I tested on. You might have to increment the CACHEBUST argument in the Dockerfile. Also, are there any outstanding edits needed? I think I fixed everything you requested earlier.

**Comment by harsha-simhadri** [comment]
Let me merge this PR since its functional. We will do a final run later and it would have your latest commit.

**Pull request merged**